include/fi_util.h

```diff
@@ -651,4 +654,43 @@ int ofi_shm_map(struct util_shm *shm, const char *name, size_t size,
 		int readonly, void **mapped);
 int ofi_shm_unmap(struct util_shm *shm);
 
+/*
+ * Name Server TODO: add support for Windows OS
+ * (osd/windows/pthread.h should be extended)
+ */
+
+enum ofi_ns_side {
+	OFI_NS_CLIENT,
+	OFI_NS_SERVER,
+}; 
+
+typedef int(*ofi_ns_service_cmp_func_t)(void *svc1, void *svc2);
+
+struct util_ns {
+	RbtHandle		ns_map;
+	pthread_t		ns_thread;
+	int			ns_port;
+	enum ofi_ns_side	ns_side;
+
+	void	*any_service_def;
+
+	size_t	name_len;
+	size_t	service_len;
+
+	ofi_ns_service_cmp_func_t	service_cmp;
+};
```

Comment:
Instead of initializing ns for "client side" and "server side", we can have an "ofi_ns_init" function that takes all the user input. The returned object can be later used for both server & client. We could even put all the user input into a "util_ns_attr" structure. A separate "ofi_ns_start" would then start the server thread. 

instead of defining a value for "any service", we can pass a function that check if a service is a wildcard, just like how the service_cmp function works. That should be more flexible than doing a memcmp.